### ISSUE TITLE
Camera extrinsics and voxel origin in NYUv2

### ISSUE THREAD

**Issue by someone-TJ** [astra-vision/MonoScene#60]
With a fixed scene size in front of the camera, should the voxel origin coordinates under the camera coordinate system be the same or approximate?

**Comment by anhquancao** [comment]
With a volume of size 6m x 4m x 2m in the (x, y, z) axis in front of the camera coordinate, the voxel origin should be: (-3m, -2m, 0). x is the horizontal axis, y is the vertical and z is the depth.

**Comment by someone-TJ** [comment]
If my understanding is correct, it means that the voxel origin of each ground truth can be set to (-3,-2,0) in the camera coordinate system?

**Comment by anhquancao** [comment]
It's the voxel origin of the 3D volume that you will predict.
You need to transform the ground-truth voxels to the camera coordinate and keep the voxels inside the volume that you defined.

**Comment by someone-TJ** [comment]
The voxel origin of each data is different after the coordinate system transformation, which causes me to still rely on reading the voxel origin of each data when training, instead of using a fixed value, which is unavoidable?

**Comment by anhquancao** [comment]
After transforming to the camera coordinate, you just keep the ones volume you defined. You remove all other.  It means that you might lose part of the ground-truth outside the volume. The network only predicts this volume in front of the camera.

**Comment by someone-TJ** [comment]
How do I use mayavi to visualize camera-view prediction in NYUv2 like this?
https://github.com/astra-vision/MonoScene/blob/master/teaser/SemKITTI.gif
I converted the voxel origin to the camer a coordinate system and commented out some code, and it looks like something went wrong，like this
![NYUv2](https://s3.bmp.ovh/imgs/2023/04/23/1728874fe12c7c8e.jpg)

**Comment by someone-TJ** [comment]
> After transforming to the camera coordinate, you just keep the voxels in the volume you defined. You remove all others. It means that you might lose part of the ground-truth outside the volume. The network only predicts the volume in front of the camera.
yes, when I get the transform done, Is it with 2.4<x<2.4,2.4<y<2.4,0<z<1.44 to pick out the voxel points inside the scene, but I can‘t reshape 60×36×60 when the matching points are less than 129600

**Comment by anhquancao** [comment]
You should create a new empty tensor and set the occupied voxels using the transformed points.
For example, assume you have the gt: pos of size (N, 3). points of size (N)
```
t = torch.zeros(60, 36, 60)
t[pos[:, 0], pos[:, 1], pos[:, 2]] = points

```

You can draw the camera as a polygon at 0, 0, 0 and the direction is the z axis.

**Comment by someone-TJ** [comment]
> After transforming to the camera coordinate, you just keep the voxels in the volume you defined. You remove all others. It means that you might lose part of the ground-truth outside the volume. The network only predicts the volume in front of the camera.
there's something confused me, I got the coordinates of the ground truth in the camera coordinate system. Assume the scene in front of camera is still (4.8,4.8,2.88) ,the vox_origin should be (-2.4,-2.4,0). what should I do?

**Comment by anhquancao** [comment]
It means that we only predict and evaluate the defined area. You can define a larger area that contains all the GT, but it will cost more memory.

**Comment by someone-TJ** [comment]
> 
This is indeed a solution, but firstly, each ground truth has a different voxel origin, which means that it is not easy to determine a uniform, larger region to contain all the ground truths, and secondly, as you said, this may introduce additional computational complexity and performance loss due to the increased proportion of empty voxels

**Comment by someone-TJ** [comment]
> It means that we only predict and evaluate the defined area. You can define a larger area that contains all the GT, but it will cost more memory.

This is indeed a solution, but firstly, each ground truth has a different voxel origin, which means that it is not easy to determine a uniform, larger region to contain all the ground truths, and secondly, as you said, this may introduce additional computational complexity and performance loss due to the increased proportion of empty voxels.

**Comment by someone-TJ** [comment]
Thank you for your previous patient reply, I have solved the problem about the camera-centric visualization, I will try other methods to remove the ground truth outside the defined volume
